### PR TITLE
Handle deprecation messages according to error_level setting

### LINKED ISSUE #3424: E_USER_DEPRECATED errors bypass error_level setting
It appears that Codeception's `Subscriber\ErrorHandler` doesn't respect the `error_level` configuration setting for `E_USER_DEPRECATION` errors.

# Background

After a recent `composer update`, one of our third-party libraries started triggering `E_USER_DEPRECATED` errors. These errors appeared as notifications after our tests. 

I tried setting Codeception's `error_level` to `E_ALL & ~E_STRICT & ~E_DEPRECATED & ~E_USER_DEPRECATED` (note the addition of `E_USER_DEPRECATED`). However, the notifications continued to appear.

config

notifications

I added `var_dump(error_reporting())` to my test suite's `_beforeSuite()` and `_afterSuite()` methods. Both methods printed `6143`, the correct value for `E_ALL & ~E_STRICT & ~E_DEPRECATED & ~E_USER_DEPRECATED`.

Error reporting appears to be correctly set from start to finish.

# Issue

The issue appears to stem from the `Subscriber\ErrorHandler`. 

As I understand it, setting PHP's error level via [`error_reporting()`](http://php.net/manual/en/function.error-reporting.php) will set the error levels for PHP's _native_ error handler.  However, when you register a _custom_ error handler via the [`set_error_handler()`](http://php.net/manual/en/function.set-error-handler.php) function, the _native_ error handlers will be bypassed for all the _custom_ handler's error types. If you don't specify error types, a _custom_ error handler defaults to `E_ALL | E_STRICT`. 

On line 36 of the `ErrorHandler` object, we correctly set the error reporting level to my configuration setting, `E_ALL & ~E_STRICT & ~E_DEPRECATED & ~E_USER_DEPRECATED`. However, on line 41, we register an error handler without specifying the error types. 

handle

Now, in the `errorHandler()` method, we do check to be sure the error being handled is one we've set with `error_reporting()`, however, this occurs after the check for `E_USER_DEPRECATED` errors. 

errorhandler

Even though I've excluded `E_USER_DEPRECATED` errors from Codeception's `error_level` reporting, they are being reported. That's not great.

# Solutions

In my mind, there are two solutions:

## Specify the handler's error types

We could specify the custom error handler's error types when we set it. This would remove the need to check the error types in the `errorHandler()` method.

solution1

## Move the E_USER_DEPRECATED check

We could move the E_USER_DEPRECATED check in the `errorHandler` method below the error types check.

solution2

Either solution fixes the issue.

success

# Pull request

The first solution seems more correct. We shouldn't handle events only to not handle them, unless we want to suppress PHP's native error handling. The second solution is definitely a much smaller change. 

I'm happy to make a pull request for either one. I wanted to check with the authors/contributors first, because there is plenty going on in here that I don't understand haha.

**Which solution do you recommend?**

### PR DESCRIPTION
**Pull request by Naktibalda**
Fixes #3424

### CODE REVIEW

**Comment by Naktibalda** [comment]
It appears that we don't want to respect error_level for all levels, it has to be limited to E_USER_DEPRECATED.

**Comment by Naktibalda** [comment]
The test is flaky, I didn't expect that.

File: tests/unit/Codeception/Subscriber/ErrorHandlerTest.php [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+
+    public function testDeprecationMessagesRespectErrorLevelSetting()
+    {
+        $errorHandler = new ErrorHandler();
+
+        $suiteEvent = new SuiteEvent(new Suite(), null, ['error_level' => 'E_ERROR']);
+        $errorHandler->handle($suiteEvent);
+
+        $messagesBeforeRun = count(Notification::all());
+        $errorHandler->errorHandler(E_USER_DEPRECATED, 'deprecated message', __FILE__, __LINE__, []);
```
Comment (Naktibalda) [inline tests/unit/Codeception/Subscriber/ErrorHandlerTest.php]: How can it add messages that happened in other tests at this line of code?

File: tests/unit/Codeception/Subscriber/ErrorHandlerTest.php [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+
+    public function testDeprecationMessagesRespectErrorLevelSetting()
+    {
+        $errorHandler = new ErrorHandler();
+
+        $suiteEvent = new SuiteEvent(new Suite(), null, ['error_level' => 'E_ERROR']);
+        $errorHandler->handle($suiteEvent);
+
+        $messagesBeforeRun = Notification::all();
+        $errorHandler->errorHandler(E_USER_DEPRECATED, 'deprecated message', __FILE__, __LINE__, []);
```
Comment (Naktibalda) [inline tests/unit/Codeception/Subscriber/ErrorHandlerTest.php]: How can it add messages that happened in other tests at this line of code?

**Comment by Naktibalda** [comment]
```
 Test  tests/unit/Codeception/Subscriber/ErrorHandlerTest.php:testDeprecationMessagesRespectErrorLevelSetting

Deprecation message was added to notifications

Failed asserting that Array &0 () is identical to Array &0 (

    0 => 'WARNING: Environments are not configured tests/data/claypit/tests/skipped/NoEnvironmentCept.php'

    1 => 'WARNING: Environments are not configured tests/data/claypit/tests/skipped/NoEnvironmentCept.php'

).

#1  /home/travis/build/Codeception/Codeception/tests/unit/Codeception/Subscriber/ErrorHandlerTest.php:21

**Comment by Naktibalda** [comment]
Oh, I haven't noticed that Notification::all() clears the old messages.

File: tests/unit/Codeception/Subscriber/ErrorHandlerTest.php [hunk 3]
```diff
@@ -0,0 +1,23 @@
+<?php
+
+use Codeception\Event\SuiteEvent;
+use Codeception\Lib\Notification;
+use Codeception\Subscriber\ErrorHandler;
+use Codeception\Suite;
+
+class ErrorHandlerTest extends \PHPUnit_Framework_TestCase
```
Comment (DavertMik) [inline tests/unit/Codeception/Subscriber/ErrorHandlerTest.php]: wow, that's cool

File: src/Codeception/Subscriber/ErrorHandler.php [hunk 4]
```diff
@@ -25,15 +25,20 @@ class ErrorHandler implements EventSubscriberInterface
...
+    {
+        $this->errorLevel = E_ALL & ~E_STRICT & ~E_DEPRECATED;
+    }
 
     public function handle(SuiteEvent $e)
     {
         $settings = $e->getSettings();
         if ($settings['error_level']) {
-            $this->errorLevel = $settings['error_level'];
+            $this->errorLevel = eval("return {$settings['error_level']};");
```
Comment (DavertMik) [inline src/Codeception/Subscriber/ErrorHandler.php]: eval 0.o
what if eval is blocked?

**Pull request merged**